### ISSUE TITLE
One keystroke lag while waiting for completion

### ISSUE THREAD

**Issue by blahgeek** [emacs-lsp/lsp-mode#2758]
So I encountered this weird problem with ts-ls server: it seems that sometimes, when emacs is waiting for the completion result, there's always one keystroke lagging. It does not happen every time, only sometimes.

For example:

---

At first, I'm in this state (`|` means the cursor): `xxx.|`. There's no completion popup yet (apparently sometimes it took ts-ls about 10 seconds to get the completion results, but I'm not complaining for that for now) and I think the ts-ls server is processing the request (CPU is 100%).

Then, I type some random key, like `a`. Nothing happens. It still displays `xxx.|`. Even after one or two seconds.

Then, I type another key, say `b`. Instantly, it displays `xxx.a|`, without `b`.

Then, I type another key, say `c`. Instantly, it displays `xxx.ab|`, without `c`.

And so on. There's always one keystroke lag. Any keys, including backspace, have the same behavior.

The lag would disappear as soon as the completion popup appears. After that, everything go back to normal.

---

My setup: emacs native-comp branch in linux. company-mode with lsp's `company-capf` as suggested.

I'm not sure what other information do you need, please let me know if you need any.

**Comment by yyoncho** [comment]
To avoid blinking of the popup lsp-mode will return the previously available results if it is unable to retrieve/calculate the new result when the user is typing. This is controlled by `lsp-completion-use-last-result`.

**Comment by blahgeek** [comment]
> To avoid blinking of the popup lsp-mode will return the previously available results if it is unable to retrieve/calculate the new result when the user is typing. This is controlled by `lsp-completion-use-last-result`.

I think it should only affect the content in the completion popup? In my case, the last character I type is not shown in the buffer. I don't think that's expected.

**Comment by yyoncho** [comment]
Got it. Which commit is your emacs built from? Can you test with other versions as well?

**Comment by blahgeek** [comment]
I'm using emacs from [nixpkgs](https://github.com/nix-community/emacs-overlay), version 20210312, which I believe is either https://github.com/emacs-mirror/emacs/commit/82bd6d57d54d4cdb205d921c2476d1dbb17f4188 or https://github.com/emacs-mirror/emacs/commit/d018584814e0c15f13bc458ba54491239b584069 (they changed the commit in the same version once so I'm not sure....)

I will update to the latest version and try again.

**Comment by blahgeek** [comment]
> I'm using emacs from [nixpkgs](https://github.com/nix-community/emacs-overlay), version 20210312, which I believe is either [emacs-mirror/emacs@82bd6d5](https://github.com/emacs-mirror/emacs/commit/82bd6d57d54d4cdb205d921c2476d1dbb17f4188) or [emacs-mirror/emacs@d018584](https://github.com/emacs-mirror/emacs/commit/d018584814e0c15f13bc458ba54491239b584069) (they changed the commit in the same version once so I'm not sure....)
> 
> I will test with latest native-comp version and 27.1 version.

I can reproduce this issue in both the latest native-comp version and 27.1 version.

**Comment by yyoncho** [comment]
Are you able to reproduce with lsp-start-plain.el + emacs 27.1? If yes, can you provide a file/project to reproduce the issue with.

**Comment by yyoncho** [comment]
(with the description from the PR I am unable to reproduce the issue).

**Comment by blahgeek** [comment]
@yyoncho  I'm able to reproduce this with lsp-start-plain.el:

1. Download `lsp-start-plain.el`, and add the following two lines:

```
(add-hook 'after-init-hook 'global-company-mode)
(setq company-idle-delay 0)
```

2. open emacs with `emacs -q -l ./lsp-start-plain.el`, install `ts-ls` language server
3. patch the `ts-ls` language server to make it slower on completion (to simulate a large project): in `.emacs.d/.cache/lsp/npm/typescript-language-server/lib/node_modules/typescript-language-server/lib/lsp-server.js` line 401 (in `completion(params)` function), add line `yield (new Promise(resolve => setTimeout(resolve, 10000)));` (always delay 10 seconds).
4. Restart emacs with the same command, open `some-new-file.js`
5. type `import`, wait 10 seconds, the autocompletion popup will apear (as expected)
6. type 6 backspaces, delete `import`, slowly start typing some other random words `abcdefg`, you will see the lag now (e.g. after typing `abcdefg`, it only displays `abcdef`

**Comment by yyoncho** [comment]
Thank you, I will take a look.

**Comment by blahgeek** [comment]
git bisect shows that this issue is introduced in https://github.com/emacs-lsp/lsp-mode/pull/2483

**Comment by blahgeek** [comment]
As I dig deeper, I think this may be a company-mode issue: I think it should trigger a redisplay before fetching candidates. If I add `(redisplay)` at the beginning of `company--fetch-candidates`, this issue will not appear.

**Comment by yyoncho** [comment]
I was suspecting the same. #2483 is fix exposing the issue. @dgutov, any thoughts?

**Comment by dgutov** [comment]
@yyoncho I don't have the code on my machine, or a reproduction scenario to fiddle with, but AFAIU this stems from the "faux async" approach you use here (also employed by @joaotavora in Eglot), that you wait for the user's next keypress before deciding what to do (show completions or abort), while the completion framework has no idea what's going on. Thus Emacs's reaction is unavoidably delayed by one keypress.

A `(redisplay)` call might be an improvement, but redisplay is not free: `(benchmark 1 '(redisplay))` shows it can add an extra 20-30ms delay, so I wouldn't want to call it unconditionally.

Try adding `(when lsp--throw-on-input (redisplay))` before the request is issued here and see if it improves things: https://github.com/emacs-lsp/lsp-mode/blob/65fb3e8d071406c4596dcc13e3f0230e1f730ec6/lsp-completion.el#L418

**Comment by yyoncho** [comment]
@dgutov makes sense.

@kiennq to me, it seems like placing `(redisplay)` just after we called the server is better since we will use the time we wait for the server's response to refresh the screen. WDYT?

FWIW redisplay on my side takes less than 1ms.

**Comment by kiennq** [comment]
> @kiennq to me, it seems like placing `(redisplay)` just after we called the server is better since we will use the time we wait for the server's response to refresh the screen. WDYT?

Yeah, that looks okay to me. Should be a change in `lsp-request-*` right?

**Comment by aaronjensen** [comment]
> FWIW redisplay on my side takes less than 1ms.

That's amazing. On my machine it takes 28ms w/ my config and anywhere from 3-20ms with `emacs -Q`

**Comment by yyoncho** [comment]
> Yeah, that looks okay to me. Should be a change in `lsp-request-*` right?

Do you mean adding it to `lsp-request` as well? I don't see a valid reason to add it to lsp-request as well but most likely it won't hurt if we do so.

**Comment by yyoncho** [comment]
> > FWIW redisplay on my side takes less than 1ms.
> 
> That's amazing. On my machine it takes 28ms w/ my config and anywhere from 3-20ms with `emacs -Q`

I believe if you are on 4k monitor(@dgutov is) it will affect redisplay speed(but I am not an expert in the field).

**Comment by aaronjensen** [comment]
Ah, I'm on a retina macbook, likely two things working against me.

**Comment by dgutov** [comment]
> I believe if you are on 4k monitor(@dgutov is) it will affect redisplay speed(but I am not an expert in the field).

It seems it's affected by whether `tool-bar-mode` is enabled (so redisplay is slow-ish with `emacs -Q`) and whether one has a fancy mode-line package installed (`smart-mode-line` in my case). The window size is more or less irrelevant.

**Comment by dgutov** [comment]
> Edit: just for clarity, we need to add it to lsp-request-while-no-input for sure

It won't affect anything else than completion, right?

**Comment by yyoncho** [comment]
> It seems it's affected by whether `tool-bar-mode` is enabled (so redisplay is slow-ish with `emacs -Q`) and whether one has a fancy mode-line package installed (`smart-mode-line` in my case). The window size is more or less irrelevant.

I don't have toolbar and I am using vanilla modeline since most of the modelines are slow...

**Comment by dgutov** [comment]
> I don't have toolbar and I am using vanilla modeline since most of the modelines are slow...

That seems like a good choice, and we could probably treat slowdown from both as bugs anyway, but until they are fixed across the most of the ecosystem, calling `redisplay` unnecessarily remains unwise.

**Comment by kiennq** [comment]
> It won't affect anything else than completion, right?

Yes. Well, we're waiting for the server response anyway, and that would be much slower than running `redisplay`

**Comment by aaronjensen** [comment]
> Yes. Well, we're waiting for the server response anyway, and that would be much slower than running `redisplay`

It's unclear that this is the case because I believe it's introducing a synchronous delay to every keypress that triggers a completion, effectively. This could have the result of making perceived latency be a bit worse while typing normally.

Also, we want to be sure to trigger the redisplay *after* making the request to the server, yes? Otherwise it just slows everything down.

**Comment by kiennq** [comment]
Unless we force it, the `redisplay` is being preempted by input. So I don't think there will be much delay perceived.

**Comment by kiennq** [comment]
PR here #2772

**Comment by joaotavora** [comment]
> "faux async" approach you use here (also employed by @joaotavora in Eglot

Faux ou vrai, it's not the same technique, it's used in more than LSP and never suffered from this problem, even with slow servers.

**Comment by dgutov** [comment]
If someone could evaluate `lsp-mode` and `eglot` side-by-side with the same LSP server in this scenario and report here, that would be great. It's possible that there is some subtlety to the latter's implementation which `lsp-mode` could adopt.

In the meantime, though, I have this report which pretty much equates lsp-mode and eglot's behavior wrt input lag: https://github.com/company-mode/company-mode/issues/1073#issuecomment-802262483

**Comment by joaotavora** [comment]
> which pretty much equates lsp-mode and eglot's behavior wrt input lag: company-mode/company-mode#1073 (comment)

To be clear, I was referring to the "one keystoke lag" as I understood it in _this_ issue.  I've never seen _that_ behaviour, but I've seen of course seen milleseconds of lag.  Emacs is single-threaded, and event-driven, much like JS and browsers. It only checks for input events, quits, etc in certain places, so depending on the operation involved (uninterruptible C-level JSON processing?) it could take a while before the user's input in seen.  Doesn't seem to be related to the "slowness of the server" (which runs in a different process anyway), but to the amount of processing that Emacs has to do to its output before it notices that it has to drop it on the floor.

> there is some subtlety to the latter's implementation which lsp-mode could adopt.

 It seems  emacs-lsp calls `accept-process-output` and `input-pending-p` repeatedly with a 0.001s separation. Seems like a lot of calls and  Elisp-level polling, but never tried it so can't speak to its merits. `jsonrpc.el` just calls `sit-for` once with 30s which calls C-level `Fread_event` with with those same 30s timeout, so waiting for events (key or input) is done in C.  Subtle or not,  quite a different technique. And `while-no-input` is yet another wholly different technique to accomplish mostly the same.

**Comment by blahgeek** [comment]
I can confirm that `eglot` does not have such issue.

> `jsonrpc.el` just calls `sit-for` once with 30s

According to the doc, `sit-for` will do redisplay before sleeping.

**Comment by dgutov** [comment]
> According to the doc, sit-for will do redisplay before sleeping.

Ah, so it's already doing what's being discussed here. Thanks for testing!

It does mean that Eglot also incurs the overhead of unnecessary redisplays (which are not free), but there doesn't seem to be any better solution without "proper" async.

**Comment by yyoncho** [comment]
@dgutov my understanding was that after each command emacs is running redisplay. If this is not the case(I am still not sure) and redisplay is heavy - do you know what is the lightweight method that runs to repainting emacs after typing a key?

**Comment by dgutov** [comment]
> my understanding was that after each command emacs is running redisplay

It does. Unless there is some pending input.

> do you know what is the lightweight method that runs to repainting emacs after typing a key?

Not really, that's just redisplay. ;-)

You end up doing it twice: if `sit-for` starts with redisplay, then aborts with user input, which results in another redisplay, you've done it twice. But if you don't do it at the beginning, that can be perceived by some users as "sluggish" (or one-keystroke-lag).

I suppose the main case that's disadvantaged is the "fast server" one: first you wait for redisplay, then you receive the response right after and redisplay again. So @blahgeek's suggestion of using a timer makes sense, but it's a balance between perceived lag and overall latency (some data about how quickly different servers respond should help).

If you do it, BTW, try a noop timer first: IIRC, Emacs automatically triggers redisplay whenever a timer fires (though that might require some additional conditions, not 100% sure).

**Comment by joaotavora** [comment]
> > According to the doc, sit-for will do redisplay before sleeping.
> Ah, so it's already doing what's being discussed here. Thanks for testing!

FWIW I've changed `sly.el` which uses the very same technique to pass `t` as the `NODISP` parameter to `sit-for`, and it doesn't seem to make any difference, I don't observe the so-called "one-keystroke-lag" behaviour at all.  I'd wager that doing the same in `jsonrpc.el` would produce similar effects.  So I'm skeptical of the conclusions being reached here.

> But if you don't do it at the beginning, that can be perceived by some users as "sluggish" (or one-keystroke-lag).

Except this one user here has tried it and doesn't see it, at least I don't see any difference when passing `nodisp` as `t` or `nil`. 

(not to mention I don't see a noticeable performance impact of `redisplay` either but that's maybe because I don't use fancy modelines or such cruft) 

Now, when the server is sluggish (i've added a 1 sec wait server side), there is indeed lag on the first keypress while waiting for the reply from the server, the reply that brings completions that reflect that keypress.  However, that lag doesn't seem to be related to any kind of 'extra redisplay'.  Does company not insert the character into the buffer before requesting the completions?  Regardless, this particular behaviour doesn't seem to be related to any kind of 'extra redisplay call' or not.

I'm still investingating though:  here's the code I was alluding to, btw: https://github.com/joaotavora/sly/blob/5966d68727898fa6130fb6bb02208f70aa8d5ce3/sly.el#L2420

**Comment by dgutov** [comment]
> Now, when the server is sluggish (i've added a 1 sec wait server side), there is indeed lag on the first keypress while waiting for the reply from the server

That's the issue under discussion (IIUC).

> (not to mention I don't see a noticeable performance impact of redisplay either but that's maybe because I don't use fancy modelines or such cruft)

Believe it or not, simply having `tool-bar-mode` enabled also counts as "fancy modeline".

> Regardless, this particular behaviour doesn't seem to be related to any kind of 'extra redisplay call' or not.

Indeed, additional redisplay (which can take up to 30ms, depending on user configuration) isn't going to cause this kind of problem. It can still be unfortunate, making Emacs a tad less responsive than it could be.

> Does company not insert the character into the buffer before requesting the completions?

Company doesn't do (or inhibit) any insertion. You're describing a problem with "missing" redisplay.

**Comment by joaotavora** [comment]
> That's the issue under discussion (IIUC).

Hmmm, I thought it was an odd off-by-one-like thing as described in the original post.  I don't see anything like that at all (but I haven't ever seen the original one)  And what's more, a user has also reported that the bug is not in Eglot.  I wonder if that user (@blahgeek  ?) can also try by adding `nodisp` as `t` to `jsonrpc.el` 's source code, recompile it, and see if it somehow exhibits the bug that it apparently not present right now.

> Company doesn't do (or inhibit) any character insertion. You're describing a problem with "missing" redisplay.

That doesn't seem to go together with your theory that Eglot and Sly are somehow insulated from this because they're using `sit-for` which redisplays, then waits.  That is because I see exactly the same behaviour regardless of passing `nodisp` as `t` or `nil` to `sit-for`.

It also doesn't seem to go together with my experiments.  I can't make it go away with some naive `redisplay` insertions.   Where exactly is that "missing" `redisplay`?  Where should I hackingly insert it to see the problem go away?

Finally, my intuition would be that it makes sense for `sit-for` to `redisplay`.  After all, if one has pressed a key and the intent is to insert it _and_ do start some background stuff, it makes some sense to show the effects of that keypress to the user while Emacs "sits" waiting for either that background stuff to finish and produce its effects or for another keypress to come in.  I don't see the inherent inefficiency of `redisplay` here, nor how non-faux "proper async" could somehow solve what seems to be a property of the problem statement itself.

> Believe it or not, simply having tool-bar-mode enabled also counts as "fancy modeline".

Interesting. I'm using a tty emacs now, and I turn that off in GUI emacs.Tool bar is pretty ugly anyway, but that's quite besides the point :sweat_smile:

**Comment by joaotavora** [comment]
> > Where should I hackingly insert it to see the problem go away?

> If I add (redisplay) at the beginning of company--fetch-candidates, this issue will not appear.

Very interesting, turns out the answer is just there, in this @blahgeek comment.  I did this and the `sly.el` lag (which is apparently _not_ the one-keystroke lag reported here) is _also_ fixed. 

In fact, looking at `company--fetch-candidates` it also explains why tweaking `NODISP` in SLY's `sit-for` has no effect: that function binds `inhibit-redisplay` to `t`. If I remove that binding `inhibit-redisplay`, then the problem is _also_ fixed.  And somehow it _still_ works if I pass `NODISP` as `t`.

So I'm leaning that `company--fetch-candidates` should either call `redisplay` itself, or at least don't inhibit the backend's redisplay attempts.  It doesn't make much difference with fast backends, but is quite a more pleasant experience with slow ones.  Is there an issue in the company repo to track this?

**Comment by joaotavora** [comment]
As a final bit of investigation, the lag I is also easily fixed outside company by simply binding `inhibit-redisplay` to `nil` around the `sit-for`. And it again works with `nodisp = t`, which suggests that some of its callees do some redisplay unconditionally (perhaps `Fread_event?`).

**Comment by dgutov** [comment]
> Very interesting, turns out the answer is just there, in this @blahgeek comment. I did this and the sly.el lag (which is apparently not the one-keystroke lag reported here) is also fixed.

Guess I should have read the full thread first. :facepalm: 

I was also operating on the assumption that the opened PR does fix the problem.

> Is there an issue in the company repo to track this?

There is an issue, yes. It links to this one, so it's visible among the comments here.

> And it again works with nodisp = t, which suggests that some of its callees do some redisplay unconditionally (perhaps Fread_event?).

I was probably the reason for the use of `inhibit-redisplay`. And `git blame` points to https://github.com/company-mode/company-mode/issues/865 as one of the reasons for that code's existence. Too bad the video is gone now.

But the description sounds like the "flickering" @blahgeek has mentioned in the PR thread.

**Comment by blahgeek** [comment]
> I can confirm that `eglot` does not have such issue.

I'm so sorry, this statement was wrong. `eglot` have the exact same issue. I got the wrong result because during the previous debugging, I've modified the `inhibit-redisplay` to `nil` in company.el, which would solve the issue for eglot.

So, essentially, it's the same behavior as described by @joaotavora :

> Now, when the server is sluggish (i've added a 1 sec wait server side), there is indeed lag on the first keypress while waiting for the reply from the server

> As a final bit of investigation, the lag I is also easily fixed outside company by simply binding inhibit-redisplay to nil around the sit-for.



> > That's the issue under discussion (IIUC).
> 
> Hmmm, I thought it was an odd off-by-one-like thing as described in the original post. I don't see anything like that at all (but I haven't ever seen the original one)

I think they are the same issue. It's just this happens for every keystroke with `lsp-mode` in some scenarios (may related to #2483), but it only happen at first (or second?) keystroke with `eglot`, hence the difference. But that's not really important for this issue.

**Comment by blahgeek** [comment]
So to summarize:

- When there's current completion available, `company-mode` may synchronously query the backend in post-command-hook after some character is inserted, which happens before normal redisplay. If the backend don't return the result very quickly, the change will not be displayed in the buffer, hence the "one-keystroke-lag" or "sluggish".
- This can be fixed by doing a manual `redisplay` somewhere during this process.
  - For `eglot`, it uses `sit-for` to wait for the input, which should already include a `redisplay`. However, `company.el` set `inhibit-redisplay` to `t` which inhibit this behavior.
  - For `lsp-mode`, there's no such mechanism now. We can add a `redisplay` before waiting the server, which is what this PR is doing https://github.com/emacs-lsp/lsp-mode/pull/2772
- However, it seems that `company-mode` would close the previous completion popup before making the new request. So always adding the `redisplay` mentioned above may introduce flickering: the popup would temporary disappear. The redisplay is unnecessary when the language server is very fast, or when in some condition, `lsp-mode` would just reuse the existing result which is also very fast.

**Comment by joaotavora** [comment]
> I think they are the same issue. It's just this happens for every keystroke with lsp-mode in some scenarios (may related to #2483), but it only happen at first (or second?) keystroke with eglot, hence the difference. But that's not really important for this issue.

If there's a difference in the number of keystrokes this effects, as you point out, it can't really be the "same" issue, can it? But there may be causes in common.

So my conclusions are different:
- there is input some lag with everything using `company.el`'s `company-capf` asynchronously.  Depending on the technique used the lag may manifest itself once (Eglot's first keypress) or multiple times (the one-keystroke-behind that @blahgeek seems to experience).
- This lag is because company is because `redisplay` shortly after a keypress isn't being called.  But it must, somehow.  How else would one see it?
- The reason it's not is being called because `company` is preventing it, the backend is failing to ensure it gets called, or both.
- In Eglot and SLY I believe this has been a problem since February 2019, where `inhibit-redisplay` was moved "higher" in `company--fetch-candidates` (company commit 2b671ecb4644b3b5714448197070ef96c67e243b).  This explains why I didn't see the problem when developing the technique in `sly.el` in 2018/2019.
- In `sly.el` and `jsonrpc.el` I'm going to bind `inhibit-redisplay` to `nil` around `sit-for`.  Not sure if I should pass `nodisp` to the `sit-for` 's
- I'd still recommend `sit-for` + `throw` + `catch` (or just `jsonrpc.el`) as the elegant technique for solving the responsiveness problem. But good luck with the timer.

I actually think that maybe `inhibit-redisplay` should stay in `company.el`: it's probably meant to prevent flickering for very fast `company-capf` backends, which maybe are the majority (but then why are these backends calling `redisplay` and making the flickering?) 

Another way is to add another `:company-thingy` key to `company-capf.el`.

**Comment by blahgeek** [comment]
@joaotavora 

> > I think they are the same issue. It's just this happens for every keystroke with lsp-mode in some scenarios (may related to #2483), but it only happen at first (or second?) keystroke with eglot, hence the difference. But that's not really important for this issue.
> 
> If there's a difference in the number of keystrokes this effects, as you point out, it can't really be the "same" issue, can it? But there may be causes in common.

Yes I think we meant the same thing. They share the same root cause.

> So my conclusions are different:
> 
>     * there is input some lag with everything using `company.el`'s `company-capf` asynchronously.  Depending on the technique used the lag may manifest itself once (Eglot's first keypress) or multiple times (the one-keystroke-behind that @blahgeek seems to experience).
> 
>     * This lag is because company is because `redisplay` shortly after a keypress isn't being called.  But it must, somehow.  How else would one see it?
> 
>     * The reason it's not is being called because `company` is preventing it, the backend is failing to ensure it gets called, or both.

Yes, I agree. I meant the same.

>     * In `sly.el` and `jsonrpc.el` I'm going to bind `inhibit-redisplay` to `nil` around `sit-for`.  Not sure if I should pass `nodisp` to the `sit-for` 's
> 
>     * I'd still recommend `sit-for` + `throw` + `catch` (or just `jsonrpc.el`) as the elegant technique for solving the responsiveness problem. But good luck with the timer.

I also support this. This is essentially also what https://github.com/emacs-lsp/lsp-mode/pull/2772 is doing for `lsp-mode` (make sure `redisplay` is called). However it may introduce flickering as I mentioned.

> I actually think that maybe `inhibit-redisplay` should stay in `company.el`: it's probably meant to prevent flickering for very fast `company-capf` backends, which maybe are the majority (but then why are these backends calling `redisplay` and making the flickering?)
> 
> Another way is to add another `:company-thingy` key to `company-capf.el`.

For this part, I have different opinion. I think the proper fix for `company-mode` is to keep the current completion result, so that the immediate `redisplay` does not introduce flickering.

**Comment by joaotavora** [comment]
> For this part, I have different opinion. I think the proper fix for company-mode is to keep the current completion result, so that the immediate redisplay does not introduce flickering.

I see.  Makes sense.  Indeed the only way to to avoid flickering but not so much as to show monstruous lag is to use a low-pass filter, in the form of a finely set timer that calls `(let ((inhibit-redisplay nil)) (redisplay))`  if it hasn't been cancelled yet.  Maybe I'll add that to `jsonrpc.el` too, or maybe it would be baked into `sit-for`, dunno.  This flickering in Eglot, doesn't seem like much of a problem to me though.  Becasue servers are generally slowish (around 0.2 - 0.5s in my case), I prefer to see the key I typed immediately being registered.

**Comment by dgutov** [comment]
@blahgeek 

> However, it seems that company-mode would close the previous completion popup before making the new request

Unfortunately, there is a technical reason for that: the overlay-based popup breaks many commands' execution unless the popup is "hidden" in `pre-command-hook` (if you try to `forward-char` while the region near point is `invisible`, the result is not what you expected). So we necessarily `hide` the overlay in `pre-command-hook` and show it again at the end of `post-command-hook`. Perhaps the conversation could be different is we could migrate to a different mechanism for all users (posframe for GUI, something else in the terminal), but that doesn't seem to be happening yet, and posframe-based frontend still have their share of glitches, as well as lower performance.

Now, we could `show` the popup again at the beginning of `company--fetch-candidates`, but that raises a few questions. First is: how to write it in a way that's useful not only for "faux async" backends? Because they take the code path for "synchronous" backends, which are supposed to return their result quickly. So the second question would be is whether it's possible to differentiate between the sync and "faux async" somehow, because re-rendering the popup, redisplaying, then doing it again after we received the results will be suboptimal. And it's not like synchronous backends are going out of style. "Fast" LSP servers could probably also be counted in that category.

"Redisplay on a timer" could be a good compromise, except the popup will still be hidden, unless `company--fetch-candidates` only re-renders it unconditionally but doesn't redisplay itself (also extra overhead, about 30-40 ms, if I'm measuring it correctly). I suppose the timer could also call `(company-call-frontends 'post-command)` itself, making that action specific to "faux async" backends. It's a hack, though, another minor goodbye to being frontend-agnostic.

> At the same time, it would be better for company-mode to keep the current completion result

`company-candidates` is updated later, so it should still have the previous value at that point.

**Comment by dgutov** [comment]
@joaotavora 

> In Eglot and SLY I believe this has been a problem since February 2019, where inhibit-redisplay was moved "higher" in company--fetch-candidates (company commit 2b671ecb4644b3b5714448197070ef96c67e243b). This explains why I didn't see the problem when developing the technique in sly.el in 2018/2019.

Just a reminder: the "problem" is the fix which you yourself suggested in the 2019 issue I linked to (invented in parallel to my own solution for flickering for async backends).

> but then why are these backends calling redisplay and making the flickering?

https://github.com/company-mode/company-mode/issues/865#issuecomment-460361590

>  Indeed the only way to to avoid flickering but not so much as to show monstruous lag is to use a low-pass filter, in the form of a finely set timer that calls (let ((inhibit-redisplay nil)) (redisplay)) if it hasn't been cancelled yet.

It's a fine intermediate solution, but without re-displaying popup the flickering is going to be there.

Here's a dirty/experimental patch that does both, but the benchmarking block reports that it takes up to 20ms:

```diff
diff --git a/company.el b/company.el
index d9d744d..26045ac 100644
--- a/company.el
+++ b/company.el
@@ -1273,13 +1273,16 @@ update if FORCE-UPDATE."
 
 (defun company--fetch-candidates (prefix)
   (let* ((non-essential (not (company-explicit-action-p)))
-         (inhibit-redisplay t)
+;         (inhibit-redisplay t)
          (c (if (or company-selection-changed
                     ;; FIXME: This is not ideal, but we have not managed to deal
                     ;; with these situations in a better way yet.
                     (company-require-match-p))
                 (company-call-backend 'candidates prefix)
               (company-call-backend-raw 'candidates prefix))))
+    (benchmark-progn
+      (when company-candidates (company-call-frontends 'post-command))
+      (redisplay))
     (if (not (eq (car c) :async))
         c
       (let ((res 'none))

```

**Comment by dgutov** [comment]
The problem does seem fundamental: either we show the newly typed character only after the (possibly long) computation is finished, or we show some old/unfinished state before that happens.

Could someone describe the approach VS Code took regarding that? I'm guessing it also shows the previous completions list until it receives the response. Or perhaps it expects that all backends do fuzzy matching, so it caches the results from the first response and does subsequent filtering on the client. And flickering is not an issue anyway before the popup is shown for the first time.

But it has to deal with "incomplete" results too, right?

One alternative approach that would simplify things is that we mandate, one way or another, that when the popup is visible (i.e. we have already received the first completion result for the given context), the subsequent requests in that completion session have to be made synchronously. Which could work if all language servers (or all major ones, at least) are smart enough to cache the original result set and just re-filter it (quickly). But any exceptions will result in poor ux.

Or we go back to the "show the previous results" option, and try to figure out a way to only do that extra redisplay when it makes sense.

**Comment by dgutov** [comment]
The latter might look like

```diff
diff --git a/company.el b/company.el
index d9d744d..afe99d0 100644
--- a/company.el
+++ b/company.el
@@ -1274,6 +1274,7 @@ update if FORCE-UPDATE."
 (defun company--fetch-candidates (prefix)
   (let* ((non-essential (not (company-explicit-action-p)))
          (inhibit-redisplay t)
+         (refresh-timer (run-with-timer 0.005 nil #'company--sneaky-refresh))
          (c (if (or company-selection-changed
                     ;; FIXME: This is not ideal, but we have not managed to deal
                     ;; with these situations in a better way yet.
@@ -1281,7 +1282,9 @@ update if FORCE-UPDATE."
                 (company-call-backend 'candidates prefix)
               (company-call-backend-raw 'candidates prefix))))
     (if (not (eq (car c) :async))
-        c
+        (progn
+          (cancel-timer refresh-timer)
+          c)
       (let ((res 'none))
         (funcall
          (cdr c)
@@ -1298,10 +1301,16 @@ update if FORCE-UPDATE."
         (while (member (car unread-command-events)
                        '(company-foo (t . company-foo)))
           (pop unread-command-events))
+        (cancel-timer refresh-timer)
         (prog1
             (and (consp res) res)
           (setq res 'exited))))))
 
+(defun company--sneaky-refresh ()
+  (when company-candidates (company-call-frontends 'post-command))
+  (let (inhibit-redisplay)
+    (redisplay)))
+
 (defun company--flyspell-workaround-p ()
   ;; https://debbugs.gnu.org/23980
   (and (bound-and-true-p flyspell-mode)
```

**Comment by joaotavora** [comment]
> Just a reminder: the "problem" is the fix which you yourself suggested in the 2019 issue I linked to (invented in parallel to 
> my own  solution for flickering for async backends).

Indeed, by that time was no longer using slow backends and didn't measure the consequences of my suggestion.  In my defense I did propose at some point "Perhaps it could delay redisplay a bit to sometimes avoid a redisplay cycle when the dialog isn't ready." .  Well that's what I'm proposing again now: delay redisplay a bit.

> company-mode/company-mode#865 (comment)  (in response to "why are the fast backends doing redisplay")

That explains the how, but doesn't explain the why. That it should be possible doens't mean it should be pervasive.

> It's a fine intermediate solution, but without re-displaying popup the flickering is going to be there.

I don't know if you followed my suggestion, which could be to keep company unchanged, and then make the backends calling sit-for be a little more careful under these rules "if you predict you're going to take a long time waiting for completions, make sure to redisplay after a short while".  I've tried that with a timer _in the backend_ and it removed flickering for very fast backends, and still shows responsiveness for for slow ones - a low pass filter.  And it's fine for all frontends.  Maybe it should be a behaviour put into `sit-for` which seems to be designed with the need for redisplaying in mind.

Of course, the same thing can be done in company, maybe it would be cleaner.  (I don't know why the extra `post-command` is for but that's fine)

Also I'd like to note that your characterization of these process-based completion backends as "faux" seems to stem from a perspective based  solely on company, where the authors have invented something they called async and treat company-capf with some underlying assumptions that don't 100% match reality.  In those situations, you say the backend is "faux", the french word for "fake", somehow associating it with illegitimacy.  But that's just your opinion.  The very same SLY backend works fine with Helm, `icomplete` and `fido-mode` for example (though inside the minibuffer), where no flickering is seen and no lag is seen.  In fact, when the new `icomplete-vertical-mode` is used, it's really visually similar to `company`.  Perhaps company could take a page from that frontend's technique.

**Comment by yyoncho** [comment]
> The problem does seem fundamental: either we show the newly typed character only after the (possibly long) computation is finished, or we show some old/unfinished state before that happens.
> 
> Could someone describe the approach VS Code took regarding that?

@dgutov see this: (in the video I press C-SPC two times to force completion dialog)

https://user-images.githubusercontent.com/13259670/115142735-06faf200-a04c-11eb-9477-4e2bcd85c926.mp4

**Comment by dgutov** [comment]
@joaotavora 

> That explains the how, but doesn't explain the why. That it should be possible doens't mean it should be pervasive.

I don't have a good answer ATM, except to say that if this scenario is realistic, we need to handle it. And that my personal config exhibits this problem as well. Does yours?

> I don't know if you followed my suggestion, which could be to keep company unchanged, and then make the backends calling sit-for be a little more careful under these rules "if you predict you're going to take a long time waiting for completions, make sure to redisplay after a short while". I've tried that with a timer in the backend and it removed flickering for very fast backends

Without also having the pseudo-popup "unhidden" before the `(redisplay)` call, this approach leads to flickering all over again in my testing, because redisplay makes popup blink.

Perhaps we're testing different scenarios, though (mine are either `company-clang` with `no-cache` enabled, or `lsp-mode` with `elixir-ls`). If you can post a self-contained example which shows the improvement you're seeing, as well as specific patches, that could help (preferably without Eglot or LSP Mode: I only have one language server installed, and it's not particularly slow or popular).

> Of course, the same thing can be done in company, maybe it would be cleaner. (I don't know why the extra post-command is for but that's fine)

That's the step which "unhides" the overlay-based popup showing the previous list of completions.

> Also I'd like to note that your characterization of these process-based completion backends as "faux" seems to stem from a perspective based solely on company, where the authors have invented something they called async and treat company-capf with some underlying assumptions that don't 100% match reality. In those situations, you say the backend is "faux", the french word for "fake", somehow associating it with illegitimacy.

These backends are clearly informed by asynchronous functions in other environments. But any async workload is a suspended or parallelized computation which, upon its creation, returns control to the caller and allows the caller to choose what to do in the case of success or failure (explicitly, without conflating outright failure with an empty list). Neither of which happens with eglot's or lsp-mode's completion tables.

As a result, for example, Company has to make educated guesses that the seemingly straightforward code can do this complicated dance with `input-pending-p` under the covers, guess when that happens using side-channels (which is only possible because I read your - eglot or lsp-mode's - code, thus breaking the abstractions) and accommodate this special case to create the best behavior.

If you have a different name for these completion tables (which are irregular, so they do need a name), I could use that.

> The very same SLY backend works fine with Helm, icomplete and fido-mode for example (though inside the minibuffer), where no flickering is seen and no lag is seen.

None of these deal with in-buffer completion. Or with the combination of idle and in-buffer completion anyway.

**Comment by dgutov** [comment]
@yyoncho

Thanks.

I can see in the video that this LS (jdt?) is slow to retrieve completions, at least at the beginning. What happens next?

I see the "Loading..." popup, is it the look of VS Code completion popup while the server has not returned any items yet? What then, it caches the response to update the popup quickly? What happens if the result set is "incomplete", which IIRC is a possibility with the LSP protocol?

What happens in the middle of the video? Did you dismiss the popup, or did it disappear because the editor had to poll the (slow) language server for completions again?

**Comment by joaotavora** [comment]
@dgutov 

> Perhaps we're testing different scenarios, though (mine are either company-clang with no-cache enabled, or lsp-mode with elixir-ls). If you can post a self-contained example which shows the improvement you're seeing, as well as specific patches, that could help (preferably without Eglot or LSP Mode: I only have one language server installed, and it's not particularly slow or popular).

Indeed ,perhaps we are talking about different things.  I'm using SLY because it's easy to hack a very slow or very fast server.  See the single commit of https://github.com/joaotavora/sly/tree/scratch/no-flicker-but-still-responsive-company. Installing SLY is reasonably straightforward, but you need a Common Lisp executable :neutral_face:.  Not very hard, let me know if you want to try it.

> But any async workload is a suspended or parallelized computation which, upon its creation, returns control to the caller and allows the caller to choose what to do in the case of success or failure

I don't understand this theorization, or how it defines a somehow more "legitimate" async.    In this specific situation, we have  to deal with a number of events: user input, a successful response, timeout errors, other timers.  All of these things can happen in unsteady rhythms and in various orders. Some, like me, would call this asynchronous, but that's just a word. The fact is that the techniques you call "faux" handle this correctly.  I'm not saying your narrow view of what "asynchronous" means can't work.  It can, and it probably does, but it (1) is incompatible with other front-ends such as `completing-read` (2) requires users to program in ways that IMO aren't as comfortable. 

As you know, if there is one thing that is fake or feigned in Emacs, it's parallelism. Emacs, like other programming models (most notably JavaScript's), is an event-based system which does one thing at a time. The techniques you call "faux" are explicitly based on this property.  I don't think they would function if this property weren't there.

> As a result, for example, Company has to make educated guesses that the seemingly straightforward code can do this complicated dance with input-pending-p under the covers, guess when that happens using side-channels (which is only possible because I read your - eglot or lsp-mode's - code, thus breaking the abstractions) and accommodate this special case to create the best behavior.

But look at `icomplete-mode`.  It doesn't do any educated guesses and works fine with SLY and "every" backend. And if you sick it at Eglot's completion table, it probably works fine too, as it's the same technique. 

>  (explicitly, without conflating outright failure with an empty list). Neither of which happens with eglot's or lsp-mode's completion tables.

Again, I can't speak for lsp-mode, but normally outright failure is signalled in the form of an Error.  Also, the primitives with "drop on floor" semantics in Jsonrpc (`jsonrpc-request`) and SLY (`sly-eval`) accept an arbitrary object to return if the request is dropped because of user input.  I haven't found a use for that feature yet, but it's there if that's somehow a turn-off, it could be use to signal some other condition, like a `quit` or some custom one.

By the way, I didn't invent these "drop on floor" semantics.  `while-no-input` is the same thing wrapped slightly differently.

> None of these deal with in-buffer completion. Or with the combination of idle and in-buffer completion anyway.

I don't see why the matter is much different.  It's an in-buffer completion where the buffer happens to be the minibuffer.  Even the technique based on overlays is similar. With `icomplete-mode` and the current SLY, there is no flickering with a fast server and neither is there unresponsive lag with slow ones. It does what could conceivably be called "the right thing" (at least in my opinion).  You seem interested in how VSCode does things, that's fine. I'm just saying you could have a look inside other parts of Emacs, too.

**Comment by yyoncho** [comment]
> @yyoncho
> 
> Thanks.
> 
> I can see in the video that this LS (jdt?) is slow to retrieve completions, at least at the beginning. What happens next?
> 
> I see the "Loading..." popup, is it the look of VS Code completion popup while the server has not returned any items yet? 

Only if you explicitly request completions. Without this, it doesn't show any indication that it is loading the completion items.

> What then, it caches the response to update the popup quickly? 

At some point, it receives a complete response and it caches it. Also, the prefix is longer, thus the completion result is retrieved much faster. In this regard, there is no difference between lsp-mode and vscode. 

> What happens if the result set is "incomplete", which IIRC is a possibility with the LSP protocol?

The behavior of vscode, when there are no completion results, is more evident when pressing backspace. 

> What happens in the middle of the video? Did you dismiss the popup, or did it disappear because the editor had to poll the (slow) language server for completions again?

So I dismiss the popup and then I request the completion for a longer prefix and start to press backspace. See how VScode is showing the old data until it retrieves the data for the shorter prefix. I think that we have discussed that in the past when lsp-mode implemented the hack which returns the old data to avoid flickering of the completion dialog. IME this is a huge UE improvement.

**Comment by dgutov** [comment]
@joaotavora 

> See the single commit of https://github.com/joaotavora/sly/tree/scratch/no-flicker-but-still-responsive-company.

Replied there.

> Some, like me, would call this asynchronous, but that's just a word. The fact is that the techniques you call "faux" handle this correctly.

To an extent. But if "asynchronous" is just a word, why complain about the phrase "faux async"? Especially without suggesting alternative names.

> requires users to program in ways that IMO aren't as comfortable.

I fail to see how they can be less comfortable than forcing every backend to re-implement this pattern in an ad-hoc fashion. It's fragile code that is not easy to get right, as evidenced also by the present discussion.

> You seem interested in how VSCode does things, that's fine.

Because I wanted to how they solved the UX problem in terms of the behavior. I've seen some people praise their responsiveness lately, in comparison to Emacs. That's orthogonal to the APIs and the implementation.

**Comment by dgutov** [comment]
@yyoncho

> So I dismiss the popup and then I request the completion for a longer prefix and start to press backspace. See how VScode is showing the old data until it retrieves the data for the shorter prefix. I think that we have discussed that in the past when lsp-mode implemented the hack which returns the old data to avoid flickering of the completion dialog. IME this is a huge UE improvement.

All right, so VS Code indeed shows "stale" data. I suppose that makes sense, if you prefer the editor to show whatever it knows as soon as it can (both the chars right after you type, and the completions as soon as they arrive). The result can be an inconsistent view, but it seems like that's unlikely to be harmful. And it moves the blame for slow responsiveness from the editor to the language server.

If someone wants to test it and give feedback, the patch in https://github.com/emacs-lsp/lsp-mode/issues/2758#issuecomment-821903370 should be a close approximation. I'll need to rewrite it, though, after some debugging.

Also, I wonder which code should be in charge of scheduling the redisplay timer. Do you want to make the choice of the timer interval inside `lsp-mode`?

**Comment by joaotavora** [comment]
> To an extent. But if "asynchronous" is just a word, why complain about the phrase "faux async"? Especially without suggesting alternative names.

As I said, it conveys the meaning of illegality or illegitimacy to one technique and purity to another, which is misleading.  As to names, I don't think it needs one.  It's just another technique in a toolbelt that a language with non-local returns allows.

> I fail to see how they can be less comfortable than forcing every backend to re-implement this pattern in an ad-hoc fashion. It's fragile code that is not easy to get right, as evidenced also by the present discussion.

That's until you get the primitives right. Which they pretty much are, except for this thing with  `company`, which can well be solved there (in fact you seem to have a patch for it already).  Anyway, if you write  JSONRPC-based backends (which are not limited to LSP/Eglot) for example, you just use the `jsonrpc-request` primitive, where the `throw/catch/sit-for` technique lives.  If your protocol isn't JSONRPC-based either (1) it could be or (2) the technique is sufficiently easy to replicate (3) some intermediate primitives could be developed to make it sufficiently easy to replicate.

As to why I think it is more comfortable, it's subjective of course, but I think the reasonably small amount of code that e.g. Eglot  employs to achieve this particular effect with mostly all frontends speaks for itself. Also, do you know "modern" Javascript? I've learned about it recently. In ES6 there is a keyword called `await` which you use with `async`  functions and is similar to `sit-for`.  Just mentioning it because that goes against your definition of that "async" is achieved by immediately returning control to the caller. It'd be nonsensical to call `await` more faux or more real than using `Promise.then()` on an immediately returned object, even though it's much more convenient in certain situations. A modern JS program will use both depending on the situation, and so should an Elisp one.  

Given the current `capf` scheme, completion against potentially slow backends is simply much better served by using the techniques you call "faux" or "ad hoc". Write once: work for `company` and everywhere else.

> I've seen some people praise their responsiveness lately, in comparison to Emacs. 

Perhaps in comparison to company, which uses `inhibit-redisplay` carelessly (apparently partly due to my careless suggestion some years ago), but perhaps not to other frontends.

**Comment by blahgeek** [comment]
> If someone wants to test it and give feedback, the patch in [#2758 (comment)](https://github.com/emacs-lsp/lsp-mode/issues/2758#issuecomment-821903370) should be a close approximation. I'll need to rewrite it, though, after some debugging.

Thanks! I gave it a try and it seems working pretty well. I will continue testing it in my daily work and will let you know if I encounter any problems.

> Also, I wonder which code should be in charge of scheduling the redisplay timer. Do you want to make the choice of the timer interval inside `lsp-mode`?

Personally I like your patch here. I think it's the right way to solve it in `company.el` instead of `lsp-mode. I don't see any downside of doing it this way.

**Comment by yyoncho** [comment]
@kiennq this is fixed after switching to `sit-for` for `lsp-request-while-no-input`, right?

**Comment by dgutov** [comment]
@yyoncho I think it still requires a patch on my side (see the previous discussion). Just give me a few more days to work on it.

**Comment by kiennq** [comment]
> @kiennq this is fixed after switching to `sit-for` for `lsp-request-while-no-input`, right?

No, it stills require patch from `company-mode`

**Comment by dgutov** [comment]
I've pushed the fix, check it out.

@blahgeek @kiennq @yyoncho

**Comment by dgutov** [comment]
Regarding "faux async", @joaotavora any anyone else who might be interested: ES6 is indeed a good reference (it is both popular and similar enough to Elisp in concurrency limitations).

`await` actually does return control to the caller (right away, or close enough). Because it's just syntactic sugar, and any function that uses `await` must be `async`: meaning, it returns a `Promise`. Check out [this guide](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Asynchronous/Async_await), for example.

It's unfortunate that we still don't have an official async convention for completion tables, for use with `icomplete` and other modes. Someday we'll get more developers with experience in asynchronous code on emacs-devel, and finally get this moving forward.

**Comment by kiennq** [comment]
@dgutov I believe we can implement similar `await/async` behavior on Emacs already with `generator.el`. There's even [`emacs-aio`](https://github.com/skeeto/emacs-aio) which provides similar `await/async` syntax.

**Comment by joaotavora** [comment]
> await actually does return control to the caller (right away, or close enough). Because it's just syntactic sugar, and any function that uses await must be async: meaning, it returns a Promise. 

@dgutov  I criticized your assertion made [here](https://github.com/emacs-lsp/lsp-mode/issues/2758#issuecomment-822069468), which I quote:

>  > > "any async workload is a suspended or parallelized computation which, upon its creation, returns control to the caller and allows the caller to choose what to do in the case of success or failure"

I did it using an example of a, different, non-faux, valid way to do async programming that doesn't conform to those suppositions of what "any" async workload is. My point is, summarily, that a primitive based on `sit-for` -- say  `jsonrpc-request` in `jsonrpc.el` --  calling is very much like calling `await jsonrpcRequest()` in JavaScript.  My point is also that this is more comfortable to the programmer.

So, either you are completely confused, or simply we are talking about totally different things.  Let me tell you what *I'm* talking about:

- _"returning control to the caller"_ : the caller calls `f()`, it executes, and if `f()` doesn't `throw` or return non-locally, the next line/statement/form after the `f()` call executes. At this point caller _knows_ that  _no_ other code (timers, process filters, acts of god other than the OS's) has intervened in the meantime.  In single-threaded Emacs and  Javascript, it's exactly the same thing.  `f()` may have scheduled a timer,  started an external process that will maybe run Elisp or Javascript.  Call this code `g()`. Also `f()` may have have returned a `Promise`, a boolean, a callback, or a hairy monkey.  Regardless, `g()` is guaranteed _not_ to have run by the time `f()` returns.  `g()` was created, but control was given immediately to `f()`. The caller of `f()` can use a variety of methods to "wait" for `g()`, most commonly it passes a callback to `f()` or does something smart with `f()`'s return value.

- _"*non*-immediate return of control to caller"_: This second model is  what a `sit-for`-powered primitive does and also what `await` in JS does. The caller calls `f()` (or `await f()` in JS). `f()` runs, schedules/creates `g()`. By now, control was **not returned** to the caller. There is no "right away, or close enough". If control is eventually handed back, many other things (timers, user input processing, etc) may have happened. By then the caller can easily know if and how `f()` and `g()` _have_ run, so it can safely continue executing based on the results of that processing. 
 
My main contention is simply that the latter is more comfortable for the programmer of RPC-style systems, such as LSP and many others. In Javascript, it avoids unnecessarily verbose `.then(function() {...})....catch()` chains.  In Emacs, it similarly avoids callback verbosity.
 
So I thank you for your guide, but I've been using async JS intensively the last 6 months (and async Elisp for much longer). I think I know how it works (at least according to my understanding of the concepts outlined above).

**Comment by joaotavora** [comment]
@kiennq  

You are right that there are other, alternatives ways to what `sit-for` accomplishes.  BTW `emacs-aio` is based on `generator.el`. 

`generator.el` a Lisp evaluator which transforms your Lisp code (which is data) into bits that can stop and resume at certain control points. It's called a CPS ("Continuation Passing Style") transformation. See [this article](http://matt.might.net/articles/cps-conversion/).  Basically, it's as if all your imperative code is quietly changed into a bunch of `Promise.then()`. As the author of `generator.el` concedes, it's slower. But that may not be a problem.   The big problem is that it won't work for every piece of pre-compiled code: that code won't be "magically" interruptible/continuable because `generator.el` has to "see" your code before applying the transformation. You **cannot** just stick a `aio-await` form in called function down the stack and expect it to work: **it won't**! `aio-await` is a macro that needs to be **lexically** inside an `iter-lamba` or `iter-defun` so that the CPS-transformator "sees" it.

Still, it might be a good thing if done correctly. Maybe `jsonrpc-request`, could, in the future, be simply implemented with `iter-defun` and `aio-await` instead of `defun` and `sit-for`.  As long as the callers of `jsonrpc-request` wouldn't notice... it's fine.

Regardless, no completion API _needs_ to be fundamentally changed or reworked in incompatible ways for that to occur, and that is a good thing.

**Comment by kiennq** [comment]
> The big problem is that it won't work for every piece of pre-compiled code: that code won't be "magically" interruptible/continuable because `generator.el` has to "see" your code before applying the transformation. You **cannot** just stick a `aio-await` form in called function down the stack and expect it to work: **it won't**! `aio-await` is a macro that needs to be **lexically** inside an `iter-lamba` or `iter-defun` so that the CPS-transformator "sees" it.

I don't see this as a big problem. When you're using `async/await` you have to follow their syntax anyway.

Pretty much everywhere requires `await` to be inside of `async` function, C++ enforces that by requiring the `co_await` has to be called inside a function that returns `std::future` or `IAsyncOperation`, C# gives compiler error when `await` is not called inside `async` function, same for Javascript.
`emacs-aio` is making the `async/await` syntax for Elisp very close with what's in the other languages, to the point that you don't have to worry about `iter-defun` anymore.

**Comment by joaotavora** [comment]
> emacs-aio is making the async/await syntax for Elisp very close with what's in the other languages, to the point that you don't have to worry about iter-defun anymore.

@kiennq  Really?? I'd be very surprised.  Can I just call `aio-await` from inside a regular function and it will just work? I think not.  Here's from the `aio-await` docstring:
```
This macro can only be used inside an async function, either
`aio-lambda' or `aio-defun'.
```

Do you want to write all your `defun` and `lambda` into slower `aio-*` versions? I think it's _a_ problem. How big, who knows? It depends.  The `sit-for` approach doesn't have this problem.

**Comment by kiennq** [comment]
@joaotavora I think you miss my part about `very close with what's in the other languages`.

Even javascript requires the `await` to be called inside `async` function so I don't see anything wrong with `aio-await` has to be called inside `aio-async`.
For write all of `defun` and `lambda` in `aio-*` form, it depends on if all of your `defun` and `lambda` are async function / coroutine or not, which I think is not remotely true. Even if it's slower, it's just an implementation detail and can be improved upon. What we gain here is similar, shorter syntax to other async/await supported language.
To call `await`, `async` inside a normal sync function, you can always use `aio-with-async` (which also similar to what other languages do).



> EDIT: `aio-defun` eventually expands to `iter-lambda`, in case the link wasn't obvious.

It's using `generator.el` so it will eventually expands to `iter-lambda`, I have no objection to that. However, it's implementation details too, and when you're using `emacs-aio`, you don't even have to write `iter-defun` or care about that.

**Comment by joaotavora** [comment]
> @joaotavora I think you miss my part about very close with what's in the other languages.

No I got it.  And I agree.  It's just that that fact, by itself, doesn't solve any problems magically.

> don't see anything wrong with aio-await has to be called inside aio-async.

There's nothing "wrong" with CPS, of course, like there is nothing wrong with programming paradigm X. That's just my point about there not being a "right" or "wrong" or "faux" way of doing things.  To each job, the most appropriate tool 

CPS transformers, whatever the language, have this characteristic: you have to rewrite the containing code where you want to call its primitives, restructuring your code using `aio-defun` and its primitives. That's some work in itself and the end result is not compatible to legacy callers who know nothing about the CPS transformer (`iter` or `aio` in Emacs's case), like completion code.  That's a drawback, a _real_ one.  In JS they rewrote wrote lots of APIs into Promise-returning API's and there are lots of adapters for old code that still works in the old style.  You know have "foo" and "fooSync" duplicates for each op in `fs`, for example. My point is that adapting old code isn't magical.

You can hide some of this complexity behind abstraction, of course.  This is why I it said `jsonrpc-request` **could** be written with `aio-await` (or maybe just `iter-next` to simplify).  But right now it's working with `sit-for`, which **also** works fine.

BTW the slowness is not an "implementation detail": you have to be careful where you use it CPS, certainly not in tight loops.  No idea how GCC Emacs fares with such code btw, but I do wonder.

**Comment by kiennq** [comment]
Also, I don't think `sit-for` is doing what similar to javascript `async/await` is doing.
I've tried this
``` elisp
(defun async-test-a ()
  (message "1")
  (sit-for 10)
  (message "2"))
(progn
  (message "before call.")
  (async-test-a)
  (message "after call."))

;; Output:
;;     before call.
;;     1
;;     2
;;     after call.
```

and this for `js`, https://jsfiddle.net/gL02dkoh/.
The output indicates that javascript `async` function returns immediately, not waiting like `sit-for`.

**Comment by kiennq** [comment]
> But right now it's working with sit-for, which also works fine.

I do agree with that, `jsonrpc-request` is working well, and trying to fix non-broken maintainable code is just non-sense for me too.

**Comment by joaotavora** [comment]
I didn't say `sit-for` is doing what `await` does, `sit-for` doesn't take a function to call, for one :wink: .  I said you can make a primitive `p()` that behaves the way `await p()` would.  That primitive is `jsonrpc-request`.  `sit-for` does what `sit-for` does :-)

> The output indicates that javascript async function returns immediately, not waiting like sit-for.

Precisely, you'd need to `await` on that second function, too. But you're in "top-level" there and top-level await is not a thing ([yet](https://v8.dev/features/top-level-await)).  You can use a IIFE for now, which is a technique used in ESM modules.
```
(async function () {
  console.log("before call.");
  await asyncCall();
  console.log("after call.");
})();
```

But this shows how in JS you have to propagate the knowledge of CPS up to the last caller.

**Comment by kiennq** [comment]
> I didn't say `sit-for` is doing what `await` does, `sit-for` doesn't take a function to call, for one 😉 . I said you can make a primitive `p()` that behaves the way `await p()` would

That would be a very limited form of `await` since `sit-for` is waiting for an interrupt from user input, whereas `await` is waiting for `p` to finish and release the control back to the caller function.

I purposely omit the `await` in the second function to illustrate that when hitting `await`, the function returns control back to the caller immediately, and the next expression will be evaluated.

**Comment by joaotavora** [comment]
> sit-for is waiting for an interrupt from user input, 

No, `sit-for` is waiting for many more things.  In particular, it's running timers, process filters and counting on a timeout.  Otherwise it wouldn't work at all!  This is why in `jsonrpc-request` you find a `catch/throw` there ;-) In fact you can say that `sit-for` is waiting for all those things + user input. 

And even if you did make a `jsonrpc-request` with fancy new CPS, I think you'd still need one or more of: `sit-for`, `while-no-input`, `accept-process-input`, `input-pending-p`.  Those are the primitives with which you'd know how and when to resolve the generator ( `iter-yield` or `aio-resolve`).

> I purposely omit the await in the second function to illustrate that when hitting await, the function returns control back to the caller immediately, and the next expression will be evaluated.

Then what did you expect? An `async` function when called _without_ `await` , returns control immediately after it's synchronous bits have run.  Which is what it did. The `console.log('2')` is _not_ one of those synchronous bits (it looks like it, but it's not, because of the CPS transformation).

**Comment by kiennq** [comment]
> In fact you can say that `sit-for` is waiting for all those things + user input.

I disagree with this. The `catch/throw` is interrupt mechanism of Elisp and `sit-for` is not waiting for those, else the `catch/throw` is not needed at all.


> Then what did you expect? An `async` function when called _without_ `await` , returns control immediately after it's synchronous bits have run. Which is what it did.

That's what make `await/async` much more powerful than just `sit-for` and also make not all primitive `p()` behaves the way `await p()` did with just `sit-for`.
For example, I have this async function that executes command and returns the output string.
``` elisp
(aio-defun test--shell (command)
  "Asynchronously execute shell command COMMAND and return its output string."
  (-let (((callback . promise) (aio-make-callback))
         (buf (generate-new-buffer " *test-shell-command*")))
    (set-process-sentinel
     (start-process-shell-command "test-shell-command" buf command)
     (lambda (proc _signal)
       (when (memq (process-status proc) '(exit signal))
         (with-current-buffer buf
           (let ((data (buffer-string)))
             (funcall callback
                      (when (> (length data) 0) (substring data 0 -1)))))
         (kill-buffer buf))))
    (car (aio-chain promise))))
```

Now can you help to turn that function into something that can be used like this using `sit-for`, not blocking and not abandoning the output either?
``` elisp
(aio-with-async
  (let (text)
    (setq text (aio-await (test--shell "ls -R /usr/")))
    (with-current-buffer (get-buffer-create "*test*")
      (insert text))))
```

**Comment by joaotavora** [comment]
> I disagree with this. The catch/throw is interrupt mechanism of Elisp and sit-for is not waiting for those, else the catch/throw is not needed at all.

Yes of course. You misunderstood me.  I meant all those things are running while `sit-for` is.  `sit-for` will only return _locally_ on timeout or user input, but many more things can happen in the background, including those that make it return non-locally (a signal, or a throw).  What I meant is that `sit-for` is not a busy loop on input only.

In your example, which is totally fine, you seem to be reinventing `async-shell-command`. But the `(aio-await (test--shell... ))` there won't return locally if the user types something.  For that you would need `sit-for` somewhere in `test--shell`. If you try to insert your code into some sync legacy code, because you that code to ask an external process for output and also to be responsive to user keystrokes (which after all is what this whole issue is about), you need `sit-for`, `while-no-input` or equivalent.  Because all the `completing-read` or `read-from-minibuffer` that Emacs interfaces has been built on for over 40 years are _not_ wrapped in `aio-with-async`.  So unless you want to tear down and rewrite all that code, it won't do what is needed. 

For new code/new interfaces that don't have these requirements, it's a very defensible idea to use `aio.el`.  I use `async/await` a lot in JS.  I _could_ have used `aio.el` when I rewrote Flymake, for example, but it wasn't (and isn't) in Emacs core and wouldn't have made such a monstrous difference anyway.  Also finding lots of `cps-state` in my backtrace is not my idea of debugging fun :-)

**Comment by dgutov** [comment]
@joaotavora 

> An async function when called without await , returns control immediately after it's synchronous bits have run.

That is exactly what I meant. Which puts the `async/await` in JS into the first category: as soon as an `await` statement in reached, and its argument is not immediate (e.g. a Promise), control gets returned to the caller function. *No functions at all* in this model are blocking.

The caller, in turn, can decide what they want to do with the returned value (which is also a Promise): they can `await` on it, or set up success/error callbacks manually, or combine several Promise values using e.g. [Promise.all](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all).

The Emacs counterpart would be returning some suspended computation and letting the caller decide what to do: whether it will `sit-for`, or `sleep`, or `accept-process-output` (thus making sure the future waits for the server response or timeouts).

The "faux async" approach in question is problematic because it takes away that choice from the caller and instead blocks, waiting for some particular condition which it has decided on its own. But it has no business making that decision.

**Comment by dgutov** [comment]
@kiennq 

`emacs-aio` is a good starting point, but it seems problematic as-is on several points:

* It provides no well-defined "promise" data type, instead defining a comfortable DSL to make one. But those functions may not always be appropriate when we want to create a promise from some existing code. Perhaps it's a bare asynchronous subprocess, perhaps it's based on a competing async API, such as [emacs-deferred](https://github.com/kiwanami/emacs-deferred).
* Another thing that's very handy to do is being able to "force" a Promise (waiting for its result synchronously). `aio`'s API suggests that we use [aio-wait-for](https://github.com/skeeto/emacs-aio/blob/master/aio.el#L159-L163), hardcoding an implementation that uses `accept-process-output` (thus being incompatible with other sources of asynchronous computations, e.g. something based on timers). I think it should be able to delegate the exact impl to the promise's creator.

I haven't tried to use it in earnest myself, though, so I could be missing something (good or bad).

**Comment by joaotavora** [comment]
> > That is exactly what I meant. Which puts the async/await in JS into the first category: as soon as an await statement in reached, and its argument is not immediate (e.g. a Promise), control gets returned to the caller function.
> as soon as an await statement in reached, and its argument is not immediate (e.g. a Promise), control gets returned to the caller function. 

This doesn't make any sense, to me.  In JS you may write (if inside an `async` context) these 3 statements involving functions `a`, `b` and `c`:

```js
a(); const retval = await b(); c(retval);
```

If `b` returns a promise, `c()` will only run only if and when that promise resolves. Lots of things may happen between that time, including things that have nothing to do with `b()` It will not run if it rejects.  Similarly in Elisp. except you don't need any special CPS context:

```lisp
(a) (setq retval (jsonrpc-request "b")) (c retval)
```

Again, lots of things can happen before `c` runs, but in both cases, from `c()` 's  point of view, it can "know" what happened of the remote request: if it finished, was interrupted, what interrupted it, errored, etc.   There is nothing "faux" or problematic about this.  It can be said that `c()`'s execution was "blocked". Maybe one of us has a bizarre understanding of what "blocking" or "synchronous" means, who knows?

Anyway this is quite in contrast to what happens if you didn't put `await` in the first example or used `jsonrpc-async-request` in the second. Also nothing "faux" or problematic about that, but `c()` has to be written quite differently. 

It turns out that most existing Elisp libs and completion-related code, `c()` is written in normal imperative without thinking about async, so it's quite more handy to be able to write to use `jsonrpc-request`. It really is, _that's_ why the problem is solved already. In another planet, if all those libs were written differently, things would be different, but as we say around my parts, if my grandmother had wheels she would be a bus.

**Comment by dgutov** [comment]
> If b returns a promise, c() will only run only if and when that promise resolves. It will not run if it rejects

Because `await` is syntactic sugar, and it makes the function be effectively rewritten in CPS under the hood to something like:

```js
a(); b().then(function (ret) {const retval = ret; c(retval);})
```

> (a) (setq retval (jsonrpc-request "b")) (c retval)

Even if it looks similar that doesn't necessarily mean it does a similar thing (it doesn't).

> There is nothing "faux" or problematic about this.

Except the things I have described in the previous message, and you snipped.

This is going nowhere again, so forgive me if I stop responding.

**Comment by joaotavora** [comment]
> Even if it looks similar that doesn't necessarily mean it does a similar thing (it doesn't).

It has the same effect, as I perfectly showed.I don't care about the implementation, having it hidden is the point! I just originally mentioned `await` so I could explain how I think it's better to use `jsonrpc-request`.  I'd probably be using `aio-await` if I could of course.  Clearly superious if you have the conditions to use it.  But we don't here, simply.

> Except the things I have described in the previous message, and you snipped.

What you said is nonsensical, so what could i say? First, the caller doesn't see `sit-for`, or `a-p-o` at all:  it sees `jsonrpc-request`. So how can I respond to that? And the caller of `jsonrpc-request` _does_ get to say what things to wait on.

> This is going nowhere again, so forgive me if I stop responding.

Don't sweat it, no prob at all, I'll sign off too. Thanks for fixing Company :+1:

**Comment by dgutov** [comment]
> it sees jsonrpc-request

The caller of `eglot-completion-at-point` does not even see that. Thus whatever modest freedoms `jsonrpc-request` does provide to its caller are irrelevant.

**Comment by yyoncho** [comment]
@dgutov I tested the latest company and it seems like the issue is gone. Marking the issue as closed.

**Comment by joaotavora** [comment]
> The caller of eglot-completion-at-point does not even see that. 

A great thing! Welcome to abstraction land!

> Thus whatever modest freedoms jsonrpc-request does provide to its caller are irrelevant.

Even better!

**Comment by wyuenho** [comment]
I'd still encountering this issue when using lsp-jedi, I'm using the latest lsp-mode and company.

**Comment by dgutov** [comment]
@wyuenho "This issue" = one keystroke lag? With the latest `company` from MELPA?

**Comment by wyuenho** [comment]
@dgutov Yes, one keystroke lag. Latest company from melpa.

**Comment by yyoncho** [comment]
(and maybe start another issue because most likely you are hitting something else).

**Comment by wyuenho** [comment]
So preliminary testing is, if I set `company-idle-delay` to 0 or something really small like 0.02, the one keystroke lag occurs very regularly. The missing character won't appear until I press another key or the server has responded to a completion request. If I leave it to the default, then it's pretty hard to trigger this bug.

A separate but perhaps related issue is, sometimes when I press 2 self-insert keys together due to typing mistakes, (say `io` together) when waiting is done, those 2 characters will be repeated 5-20 times. This is happening regardless of what I set `company-idle-delay` to.

**Comment by joaotavora** [comment]
If it helps, after switching to the latest `company.el` I don't notice any lag anymore, in any of my two extensions, Eglot and SLY, that use Company. Previously, they suffered from a lag-related problem similar (but not quite as serious) to the one described in this issue's title.  Emacs 28 and 27

**Comment by dgutov** [comment]
You can also experiment with different values of `company-async-redisplay-delay`. E.g. setting it to 0 or 0.001.
